File: rpc/response/Response.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.rpc.response;

import grakn.client.common.exception.GraknClientException;
import grakn.protocol.TransactionProto;
import io.grpc.StatusRuntimeException;

import javax.annotation.Nullable;
import java.util.Objects;

import static grakn.common.util.Objects.className;

public class Response {

    private final TransactionProto.Transaction.Res response;
    private final StatusRuntimeException error;

    private Response(@Nullable TransactionProto.Transaction.Res response, @Nullable StatusRuntimeException error) {
        if (!(response == null || error == null)) {
            throw new GraknClientException(new IllegalArgumentException("One of Transaction.Res or StatusRuntimeException must be null"));
        }
        this.response = response;
        this.error = error;
    }

    static Response completed() {
        return new Response(null, null);
    }

    static Response error(StatusRuntimeException error) {
        return new Response(null, error);
    }

    static Response ok(TransactionProto.Transaction.Res response) {
        return new Response(response, null);
    }

    @Nullable
    public TransactionProto.Transaction.Res response() {
        return response;
    }

    public final Type type() {
        if (response != null) {
            return Type.OK;
        } else if (error != null) {
            return Type.ERROR;
        } else {
            return Type.COMPLETED;
        }
    }

    public final TransactionProto.Transaction.Res ok() {
        if (response != null) {
            return response;
        } else if (error != null) {
            // TODO: parse different GRPC errors into specific GraknClientException
            throw new GraknClientException(error);
        } else {
            throw new GraknClientException("Transaction interrupted, all running queries have been stopped.");
        }
    }

    @Override
    public String toString() {
        return className(getClass()) + "{response=" + response + ", error=" + error + "}";
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (o instanceof Response) {
            final Response that = (Response) o;
            return Objects.equals(this.response, that.response)
                    && Objects.equals(this.error, that.error);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(response, error);
    }

    public enum Type {
        OK, ERROR, COMPLETED
    }
}

File: http-ts/src/response.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Database, User } from "./index";
import { Concept } from "./concept";
import { QueryStructure } from "./query-structure";

export interface SignInResponse {
    token: string;
}

export type Distribution = `TypeDB Cluster` | `TypeDB CE`;

export interface VersionResponse {
    distribution: Distribution;
    version: string;
}

export interface DatabasesListResponse {
    databases: Database[];
}

export interface UsersListResponse {
    users: User[];
}

export interface TransactionOpenResponse {
    transactionId: string;
}

export type QueryType = "read" | "write" | "schema";

export type AnswerType = "ok" | "conceptRows" | "conceptDocuments";

export interface ConceptRow {
    [varName: string]: Concept | undefined;
}

export interface ConceptRowAnswer {
    involvedBlocks: number[];
    data: ConceptRow;
}

export type ConceptDocument = Object;

export type Answer = ConceptRowAnswer | ConceptDocument;

export interface QueryResponseBase {
    answerType: AnswerType;
    queryType: QueryType;
    comment: string | null;
    query: QueryStructure | null;
}

export interface OkQueryResponse extends QueryResponseBase {
    answerType: "ok";
}

export interface ConceptRowsQueryResponse extends QueryResponseBase {
    answerType: "conceptRows";
    answers: ConceptRowAnswer[];
}

export interface ConceptDocumentsQueryResponse extends QueryResponseBase {
    answerType: "conceptDocuments";
    answers: ConceptDocument[];
}

export type QueryResponse = OkQueryResponse | ConceptRowsQueryResponse | ConceptDocumentsQueryResponse;

export type ApiOkResponse<OK_RES = {}> = { ok: OK_RES };

export type ApiError = { code: string; message: string };

export interface ApiErrorResponse {
    err: ApiError;
    status: number;
}

export function isApiError(err: any): err is ApiError {
    return err != null && typeof err.code === "string" && typeof err.message === "string";
}

export type ApiResponse<OK_RES = {} | null> = ApiOkResponse<OK_RES> | ApiErrorResponse;

export function isOkResponse<OK_RES>(res: ApiResponse<OK_RES>): res is ApiOkResponse<OK_RES> {
    return "ok" in res;
}

export function isApiErrorResponse(res: ApiResponse): res is ApiErrorResponse {
    return "err" in res;
}
